services/imagery/src/image-store.js

```diff
@@ -4,164 +4,270 @@ import path from 'path';
+   * images included.
+   *
+   * Returns the id number for the image (the first one is 1).
+   *
+   * @param  {Buffer}        image
+   * @param  {imagery.Image} metadata
+   * @param  {number}        id
+   * @return {Promise.<number>} The id number for the image.
+   */
+  async addImage(image, metadata, id = undefined) {
+    let db = await this._dbPool.acquire();
+
+    // Allow only one image at a time to be added to the database.
+    await db.run('BEGIN TRANSACTION');
+
+    try {
+      if (id === undefined) {
+        await db.run('INSERT INTO images DEFAULT VALUES');
+        id = db.lastID;
+      } else {
+        await db.run('INSERT INTO images VALUES (?)', id);
+      }
+
+      // Set the id number in the metadata.
+      metadata.id = id;
+
+      await this.setImage(id, image);
+      await this.setMetadata(id, metadata);
+
+      // Adding this to the list for rate calculations.
+      this._recordImageTime();
+
+      // Broadcast the new image id.
+      this.emit('image', id);
+
+      await db.exec('COMMIT');
+
+      await this.purgeImages();
+
+      return id;
+    } catch (e) {
+      await db.exec('ROLLBACK');
+      throw e;
+    } finally {
+      this._dbPool.release(db);
     }
+  }
 
-    /** Get the number of images stored. */
-    getCount() {
-        return this._count;
-    }
+  /** Remove old images such that the image store is no longer
+      above the limit. */
+  async purgeImages() {
+    if (!this._maxImages) return;
 
-    /**
-     * Feed an image into the image store.
-     *
-     * The metadata attached is the Image proto message without the
-     * images included.
-     *
-     * Returns the id number for the image (the first one is 0).
-     *
-     * @param  {Buffer}        image
-     * @param  {imagery.Image} metadata
-     * @return {Promise.<number>} The id number for the image.
-     */
-    async addImage(image, metadata) {
-        let id = this._count;
-
-        // Set the id number in the metadata.
-        metadata.id = id;
-
-        await this.setImage(id, image);
-        await this.setMetadata(id, metadata);
-
-        // Recording the count in case the image store restarts.
-        await fs.writeFile(COUNT_FILE, JSON.stringify(
-            { count: id + 1 }, null, 2
-        ));
-
-        // Adding this to the list for rate calculations.
-        this._recordImageTime();
-
-        // The count in incremented towards the end, to prevent image
-        // requests while still writing them.
-        this._count++;
-
-        // Broadcast the new image id.
-        this.emit('image', id);
-
-        return id;
-    }
+    let db = await this._dbPool.acquire();
 
-    /** Return the image for the id. */
-    async getImage(id) {
-        let filename = this._formatFilename(id);
+    await db.run('BEGIN TRANSACTION');
 
-        return await fs.readFile(filename, { encoding: null });
-    }
+    try {
+      while (await this.getCount() > this._maxImages) {
```

Comment:
Do not use `getCount` - make the request directly, since `getCount` will grab a new connection from the pool and therefore might block on our transaction, resulting in a deadlock. Cache the count just before the `while` - we don't need to be making the request constantly because we have exclusive access to the database while the transaction is being composed.